rfcs/12-primitive-shapes.md

```diff
@@ -0,0 +1,514 @@
+
+### Where are the `Transform`s?
+
+ Translation and rotation are **not** defined using bevy's `Transform` components. This is because types that implement `Bounding` and `Collider` must be fully self-contained. This:
+ 
+ * makes the API simpler when using these components in functions and systems
+ * ensures bounding and collision types use an absolute minimum of memory
+ * prevents errors caused by nonuniform scale invalidating the shape of the primitive.
+
+Instead, changes to the parent's `GlobalTransform` should be used to derive a new primitive on scale changes - as well as a new translation and rotation if required.
+
+### Meshing
+
+Both 2d and 3d primitives can implement the `Meshable` trait to provide the ability to generate a tri mesh:
+
+```rust
+let circle_mesh: Mesh = Circle2d{ radius: 2.0 }.mesh();
+```
+The base primitive types only define the shape (`Circle2d`) and size (`radius`) of the geometry about the origin. Once generated, the mesh can have a transform applied to it like any other mesh, and it is no longer tied to the primitive that generated it. The `Default::default()` implementation of primitives should be a "unit" variant, such as a circle with diameter 1.0, or a box with all edges of length 1.0.
+
+Meshing could be naturally extended with other libraries or parameters. For example, a sphere by default might use `Icosphere`, but could be extended to generate UV spheres or quad spheres. For 2d types, we can use a crate such as `lyon` to generate 2d meshes from parameterized primitive shapes within the `Meshable` interface.
+
+### Bounding
+
+For use in spatial queries, broad phase collision detection, and raycasting, `Bounding` and `Bounding2d` traits are provided. These traits are implemented for types that define position in addition to the shape and size defined in the base primitive. The basic functionality of this trait is to check whether one bounding shape is contained within another bounding shape.
+
+### Colliders
+
+Colliders can provide intersection checks against other colliders, as well as check if they are within a bounding volume.
+
+### 3D and 2D
+
+This RFC provides independent 2d and 3d primitives. Recall that the purpose of this is to provide lightweight types, so there are what appear to be duplicates in 2d and 3d, such as `Line` and `Line2d`. Note that the 2d version of a line is smaller than its 3d counterpart because it is only defined in 2d. 3d geometry (or 2d with depth) is assumed to be the default for most cases. The names of the types were chosen with this in mind.
+
+## Implementation strategy
+
+### Helper Types
+
+```rust
+/// Stores an angle in radians, and supplies builder functions to prevent errors (from_radians, from_degrees)
+struct Angle(f32);
```

Comment:
I initially had an `Angle` type in glam, but moved away from it. Embark raised an issue asking if it to be removed,  I was also thinking about getting rid of it myself and obliged. All trig functions in the Rust standard library work in radians so it seemed reasonable. I guess in generally I minimised the types involved in glam, there's a downside in that types communicate information and can prevent some invalid usage but they also can get in the way a bit.

If you are adding newtypes I would recommend thinking about what operations they might have, as it could increase scope and complexity of the feature.

Reply:
I would **strongly** advocate for a strongly typed `Angle` struct and personally discourage using `f32/64` as angles.
One of the primary reasons to use Rust is to utilize it's type system. Using the type system helps prevent a huge class of errors (such as mixing degrees/radians). 
Sure they can get in a the way a little bit, however I would prefer something with a minor inconvenience but strong safety guarantees over something super simple but that is super easy to misuse.

Reply:
However yes, it is very important to provide a fleshed out API to see how's these types would actually be used inside code.

Reply:
My intent here is, as @NathanSWard said, is to use the type system to communicate the type of angle and prevent this class of errors. Newtyping also makes it trivial to add helper functions like `from_degrees()` to hopefully offset the friction that adding a newtype might have.

What's missing from the current iteration of this RFC were some arcs that used this angle, though I am using it in the `RegularPolygon`. I've been struggling a bit to manage the scope of this RFC. On one hand, I want to provide types we can use everywhere in Bevy, however I don't want to specify their usage in all these places and turn this RFC into something completely unmanageable.

Reply:
Are we sure we want to use an underlying f32 value for angles? That will make reversible operations quite iffy, which can have serious gameplay (or especially CAD) impacts if you want to reset to or align to a fixed angle.

Reply:
So, from the limited CAD-engine side of things I've done, I think it's best to handle that on the business logic (or CAD engine) side, not in Bevy. One of the stated goals of this RFC is to keep the types very lightweight. The only reason I included the `Angle` type is because it is ultimately zero-cost: it gives users some typed compile time constraints without having any impact on performance.

If you need precision and reversibility, then you probably want to store precise representations in a specialized library, operate on that data there, then cross the crate boundary and convert to the Bevy types for rendering.

Reply:
Ultimately I'm of the opinion that angles in degrees have no place in a game engine outside of the editor UI and editor data formats so I guess that colours my veiwpoint a bit :) Not a hill I would die on though.

Reply:
Angles in degrees are much better for reversibility though ;)

Reply:
I'm with @bitshifter on this one :)  I'm not really convinced that ⅐ of a turn is materially more reversible in degrees than in radians.

That said, I still like strongly-type angle types, because I'm a fan of having `1.0` represent π radians.  There's even [`sinpi`, `cospi`, and friends in IEEE754](https://en.wikipedia.org/wiki/IEEE_754#Recommended_operations) to make it easy to work with that -- and they might help alice reverse things, too, since `cospi(acospi(0))` is exactly zero, whereas `f64::acos(0.0).cos() = 0.00000000000000006123233995736766`.